File: Back/app/business/services/bigquery_service.py
from typing import List, Optional
from app.integration.repositories.bigquery_repository import BigQueryRepository
from app.business.models.schemas import ConflictSummary, CountryStats

class BigQueryService:
    """Servicio de lógica de negocio para conflictos"""
    
    def __init__(self):
        self.repository = BigQueryRepository()
    
    def check_connection(self) -> bool:
        """Verificar conexión a BigQuery"""
        return self.repository.check_connection()
    
    def get_all_conflicts(
        self,
        limit: int = 100,
        offset: int = 0,
        country: Optional[str] = None,
        region: Optional[str] = None
    ) -> tuple[List[ConflictSummary], int]:
        """
        Obtener conflictos con lógica de negocio aplicada
        """
        # datos del repositorio
        raw_conflicts, total = self.repository.find_all(
            limit=limit,
            offset=offset,
            country=country,
            region=region
        )
        
        # Transformar a modelos de dominio
        conflicts = [self._dict_to_conflict(row) for row in raw_conflicts]
        
        return conflicts, total
    
    def get_conflict_by_id(self, event_id: str) -> Optional[ConflictSummary]:
        """Obtener conflicto por ID"""
        raw_conflict = self.repository.find_by_id(event_id)
        
        if raw_conflict:
            return self._dict_to_conflict(raw_conflict)
        return None
    
    def get_top_countries(self, limit: int = 10) -> List[CountryStats]:
        """Top países con transformación a modelo de dominio"""
        raw_stats = self.repository.find_top_countries(limit)
        
        stats = []
        for row in raw_stats:
            stats.append(CountryStats(
                country=row["country"],
                total_events=row["total_events"],
                total_deaths=int(row["total_deaths"] or 0),
                conflicts=list(row["conflicts"]) if row.get("conflicts") else []
            ))
        
        return stats
    
    def get_total_stats(self) -> dict:
        """Estadísticas globales"""
        raw_stats = self.repository.get_global_stats()
        
        return {
            "total_events": raw_stats["total_events"],
            "total_deaths": int(raw_stats["total_deaths"] or 0),
            "total_countries": raw_stats["total_countries"],
            "total_conflicts": raw_stats["total_conflicts"]
        }
    
    def search_conflicts(
        self,
        search_term: str,
        limit: int = 20
    ) -> List[ConflictSummary]:
        """Buscar conflictos"""
        raw_conflicts = self.repository.search(search_term, limit)
        return [self._dict_to_conflict(row) for row in raw_conflicts]
    
    def _dict_to_conflict(self, row: dict) -> ConflictSummary:
        """Transformar dict a modelo de dominio"""
        return ConflictSummary(
            event_id=row["event_id"],
            conflict_name=row["conflict_name"],
            type_of_violence=row["type_of_violence"],
            side_a=row["side_a"],
            side_b=row["side_b"],
            country=row["country"],
            region=row["region"],
            date_start=row["date_start"],
            date_end=row["date_end"],
            deaths_a=row["deaths_a"],
            deaths_b=row["deaths_b"],
            deaths_total=row["deaths_total"],
            length_of_conflict=row["length_of_conflict"],
        )

File: Frontend/angular-frontend/src/app/core/services/bigquery.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../../environments/environment';

export interface ConflictSummary {
  event_id: string;
  conflict_name: string;
  type_of_violence: string;
  side_a: string;
  side_b: string;
  country: string;
  region: string;
  date_start: string;
  date_end: string;
  deaths_a: number;
  deaths_b: number;
  deaths_total: number;
  length_of_conflict: number;
}

export interface ConflictListResponse {
  total: number;
  conflicts: ConflictSummary[];
}

export interface CountryStats {
  country: string;
  total_events: number;
  total_deaths: number;
  conflicts: string[];
}

export interface StatsResponse {
  top_countries: CountryStats[];
  total_events: number;
  total_deaths: number;
}

@Injectable({
  providedIn: 'root',
})
export class BigqueryService {
  private baseUrl = `${environment.apiUrl}/bigquery`;

  constructor(private http: HttpClient) {}

  /**
   * Obtener lista de conflictos con paginación
   */
  getConflicts(
    limit: number = 100,
    offset: number = 0,
    country?: string,
    region?: string
  ): Observable<ConflictListResponse> {
    let params = new HttpParams()
      .set('limit', limit.toString())
      .set('offset', offset.toString());

    if (country) params = params.set('country', country);
    if (region) params = params.set('region', region);

    return this.http.get<ConflictListResponse>(`${this.baseUrl}/conflicts`, {
      params,
    });
  }

  /**
   * Obtener conflicto por ID
   */
  getConflictById(eventId: string): Observable<ConflictSummary> {
    return this.http.get<ConflictSummary>(
      `${this.baseUrl}/conflicts/${eventId}`
    );
  }

  /**
   * Obtener estadísticas generales
   */
  getStatistics(topCountries: number = 10): Observable<StatsResponse> {
    const params = new HttpParams().set(
      'top_countries',
      topCountries.toString()
    );
    return this.http.get<StatsResponse>(`${this.baseUrl}/stats`, { params });
  }

  /**
   * Buscar conflictos
   */
  searchConflicts(
    query: string,
    limit: number = 20
  ): Observable<ConflictSummary[]> {
    const params = new HttpParams()
      .set('q', query)
      .set('limit', limit.toString());

    return this.http.get<ConflictSummary[]>(`${this.baseUrl}/search`, {
      params,
    });
  }

  /**
   * Obtener top países
   */
  getTopCountries(limit: number = 10): Observable<CountryStats[]> {
    const params = new HttpParams().set('limit', limit.toString());
    return this.http.get<CountryStats[]>(`${this.baseUrl}/countries/top`, {
      params,
    });
  }
}
